src/Main.elm

```diff
@@ -1,10 +1,418 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser
+import Data.Api as Api
+import Data.Audience as Audience exposing (Audience, AudienceType)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.Store as Store exposing (AudienceFolderID, AudienceLevel, Store)
+import Element exposing (Element, column, el, none, paragraph, row, text)
+import Element.Background
+import Element.Border
+import Element.Events
+import Element.Font
+import Element.Input exposing (button)
+import Html
+import Html.Attributes
+import Http
+import Task
 
 
-main : Html msg
+
+---------------
+-- M O D E L --
+---------------
+
+
+type alias InitialisingState =
+    { folders : Maybe (List AudienceFolder)
+    , audiences : Maybe (List Audience)
+    }
+
+
+type Location
+    = Root
+    | Folder AudienceFolderID
+
+
+type alias SucceedState =
+    { location : Location
+    , filterBy : AudienceType
+    }
+
+
+type Model
+    = Initialising InitialisingState
+    | Failed Http.Error
+    | Succeed Store SucceedState
+
+
+makeSelector : SucceedState -> Store.Selector
+makeSelector state =
+    case state.filterBy of
+        Audience.Authored ->
+            case state.location of
+                Root ->
+                    Store.OnlyAuthored
+
+                Folder folderID ->
+                    Store.OnlyAuthoredIn folderID
+
+        Audience.Shared ->
+            Store.OnlyShared
+
+        Audience.Curated ->
+            case state.location of
+                Root ->
+                    Store.OnlyCurated
+
+                Folder folderID ->
+                    Store.OnlyCuratedIn folderID
+
+
+turnInitialising : InitialisingState -> Model
+turnInitialising state =
+    case Maybe.map2 Store.init state.folders state.audiences of
+        Nothing ->
+            Initialising state
+
+        Just store ->
+            Succeed store (SucceedState Root Audience.Authored)
+
+
+init : flags -> ( Model, Cmd Msg )
+init _ =
+    ( Initialising
+        { folders = Nothing
+        , audiences = Nothing
+        }
+    , Cmd.batch
+        [ Task.attempt LoadAudienceFoldersDone Api.getListOfAudienceFolders
+        , Task.attempt LoadAudiencesDone Api.getListOfAudiences
+        ]
+    )
+
+
+
+-----------------
+-- U P D A T E --
+-----------------
+
+
+type Msg
+    = LoadAudienceFoldersDone (Result Http.Error (List AudienceFolder))
+    | LoadAudiencesDone (Result Http.Error (List Audience))
+    | GoDown AudienceFolderID
+    | GoUp
+    | SetFilterBy AudienceType
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case ( msg, model ) of
```

Comment:
I'd prefer having multiple `update` functions - one for each `Model` case ... less stuff to think about

Reply:
That `update` is too small to be divided I think.

Reply:
I've dave a little bit deeper and remembered why I'm using the composite `case (msg, model) of` approach.

Let's say we have `Model` and `Msg`. Each of the variants must be handled according postfix (`ModelA` with `MsgA`, etc...).

```elm
type Model
    = ModelA Int
    | ModelB Int
    | ModelC Int


type Msg
    = MsgA Int
    | MsgB Int
    | MsgC Int
```

The composite function will look like this:

```elm
update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            ModelB (a + x)

        ( MsgA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            ModelC (b * x)

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            ModelA (c - x)

        ( MsgC _, _ ) ->
            model
```

It prevents you from not handling new variants. Let's say we add `MsgAA` to handle when `ModelA`:


```elm
type Msg
    = MsgA Int
    | MsgAA
    | MsgB Int
    | MsgC Int

-- compiling error
```

Compiler will say us to handle the case:

```elm
update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            ModelB (a + x)

        ( MsgA _, _ ) ->
            model

        ( MsgAA, ModelA x ) ->
            ModelB x

        ( MsgAA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            ModelC (b * x)

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            ModelA (c - x)

        ( MsgC _, _ ) ->
            model
```

But let's separate the `update` to multiple functions:

```elm
updateA : Msg -> Int -> Model
updateA msg x =
    case msg of
        MsgA a ->
            ModelB (a + x)

        _ ->
            ModelA x


updateB : Msg -> Int -> Model
updateB msg x =
    case msg of
        MsgB b ->
            ModelC (b * x)

        _ ->
            ModelB x


updateC : Msg -> Int -> Model
updateC msg x =
    case msg of
        MsgC c ->
            ModelA (c - x)

        _ ->
            ModelC x
```


The problem is that when I add one more Msg for ModelA the compiler will say me nothing but the case will not be handled.

```elm
type Msg
    = MsgA Int
    | MsgAA
    | MsgB Int
    | MsgC Int

-- no more changes but still compiling
```

To prevent this situation I could handle all of the cases directly:

```elm
updateA : Msg -> Int -> Model
updateA msg x =
    case msg of
        MsgA a ->
            ModelB (a + x)

        MsgB _ ->
            ModelA x

        MsgC _ ->
            ModelA x


updateB : Msg -> Int -> Model
updateB msg x =
    case msg of
        MsgA _ ->
            ModelB x

        MsgB b ->
            ModelC (b * x)

        MsgC _ ->
            ModelB x


updateC : Msg -> Int -> Model
updateC msg x =
    case msg of
        MsgA _ ->
            ModelC x

        MsgB _ ->
            ModelC x

        MsgC c ->
            ModelA (c - x)
```

As you can see the previous problem is solved, but another created: you mast describe too much.

Otherwise composite approach prevent you from both of the issues. `update` is becoming bigger and bigger but instead of splitting by model you could split by cases, like:


```elm
hugeACase : Int -> Int -> Model
hugeACase a x =
    ModelB (a + x)


hugeBCase : Int -> Int -> Model
hugeBCase b x =
    ModelC (b * x)


hugeCCase : Int -> Int -> Model
hugeCCase c x =
    ModelC (c - x)


update : Msg -> Model -> Model
update msg model =
    case ( msg, model ) of
        ( MsgA a, ModelA x ) ->
            hugeACase a x

        ( MsgA _, _ ) ->
            model

        ( MsgB b, ModelB x ) ->
            hugeBCase b x

        ( MsgB _, _ ) ->
            model

        ( MsgC c, ModelC x ) ->
            hugeCCase

        ( MsgC _, _ ) ->
            model
```

Reply:
Basically for me this boils into "in one big update you can go by Msgs first, Models second - you handle all Msgs and it doesn't matter if you wildcard Models" ... compared to "separate updates for each Model would have to handle all Msgs too - more boilerplate" ... Fair enough!